torch/_inductor/memory.py

```diff
@@ -0,0 +1,555 @@
+            ):
+                name_to_input_buf[dep.name] = FreeableInputBuffer(dep)
+                name_to_input_buf[dep.name].mpi_buffer.size_free = dep_size_hint(dep)
+
+    return name_to_input_buf
+
+
+def compute_size_for_scheduler_buffer(name_to_buf: Dict[str, SchedulerBuffer]) -> None:
+    """
+    Compute the size of each scheduler buffer, including (1) memory allocated when
+    it is created and (2) memory deallocated when it is freed.
+
+    We specially handle the case of MultiOutputLayout.
+    Consider the following case:
+        buf0 = some_ops_with_multi_outputs(...)
+        buf1 = buf0[0] # assume 10 bytes
+        buf2 = buf0[1] # assume 20 bytes
+    In such cases,
+        buf0: at creation, 30 bytes allocated, when deleted, 0 bytes freed
+        buf1: at creation, 0 bytes allocated, when deleted, 10 bytes freed
+        buf2: at creation, 0 bytes allocated, when deleted, 20 bytes freed
+    """
+    from .scheduler import BaseSchedulerNode, OutputNode
+
+    # compute the size of SchedulerBuffer without MultiOutputLayout layout
+    for sched_buf in name_to_buf.values():
+        if not isinstance(sched_buf.node.layout, MultiOutputLayout):
+            sched_buf.mpi_buffer = MemoryPlanningInfoForBuffer()
+            sched_buf.mpi_buffer.size_alloc = V.graph.sizevars.size_hint(
+                sched_buf.node.get_numel(), fallback=0
+            ) * get_dtype_size(sched_buf.node.get_dtype())
+            sched_buf.mpi_buffer.size_free = sched_buf.mpi_buffer.size_alloc
+
+    # compute the size of SchedulerBuffer with MultiOutputLayout layout
+    for sched_buf in name_to_buf.values():
+        if isinstance(sched_buf.node.layout, MultiOutputLayout):
+            sched_buf.mpi_buffer = MemoryPlanningInfoForBuffer()
+            for user in sched_buf.users:
+                if isinstance(user.node, OutputNode):
+                    continue
+                assert isinstance(user.node, BaseSchedulerNode)
+                for buf in user.node.get_outputs():
+                    sched_buf.mpi_buffer.size_alloc += buf.mpi_buffer.size_alloc
+                    buf.mpi_buffer.size_alloc = 0
+
+
+def map_successor_nodes_with_predecessor_buffers(
+    nodes: List[BaseSchedulerNode],
+    name_to_input_buf: Dict[str, FreeableInputBuffer],
+    name_to_buf: Dict[str, SchedulerBuffer],
+) -> None:
+    """
+    For scheduling and memory estimation, for each scheduler node, we maintain
+    a list of its dependency buffers (SchedulerBuffer and FreeableInputBuffer).
+    This is similar to node.read_writes.reads, which is a list of Dep.
+    Reversely, for each SchedulerBuffer / FreeableInputBuffer, assign its successor nodes.
+    A buffer's successor nodes determines when a buffer can be freed.
+    """
+    for node in nodes:
+        node.mpi_node = MemoryPlanningInfoForNode()
+        node.mpi_node.pred_buffers = []
+        for dep_name in {dep.name for dep in node.unmet_dependencies}:
```

Comment:
I used `sorted()` here. To add a bit more context. I used set comprehension here because I noticed that sometimes the different `dep`s could have the same time, I guess due to aliasing. (adding a comment here as well)

> I need to add a linter 

haha.